### PR TITLE
[SYCL] vec abi changes behind -fpreview-breaking-changes flag

### PR DESCRIPTION
**Pull request by cperkinsintel**
sycl::vec was updated earlier to be SYCL 2020 conformant. But those were ABI breaking changes, and the decision has been made to put those changes behind the -fpreview-breaking-changes / __INTEL_PREVIEW_BREAKING_CHANGES  flag/macro.

This PR restores the old sycl::vec implementation and puts the new conforming one behind the flag/macro.

### CODE REVIEW

File: sycl/include/sycl/buffer.hpp [hunk 1]
```diff
@@ -168,8 +168,10 @@ template <typename T, int dimensions = 1,
               typename std::enable_if_t<(dimensions > 0) && (dimensions <= 3)>>
 class buffer : public detail::buffer_plain,
                public detail::OwnerLessBase<buffer<T, dimensions, AllocatorT>> {
+#ifdef __INTEL_PREVIEW_BREAKING_CHANGES
```
Comment (aelovikov-intel) [inline sycl/include/sycl/buffer.hpp:171]: Why this change? If it's ultimately needed than we probably need an `#else` with the assert prior to the current one.

File: sycl/include/sycl/types.hpp [hunk 2]
```diff
@@ -58,6 +75,20 @@
...
 #include <variant>     // for tuple, variant
 
+#ifndef __INTEL_PREVIEW_BREAKING_CHANGES
+
+#ifndef __SYCL_USE_EXT_VECTOR_TYPE__
+#include <sycl/detail/cl.h>
+#endif
+
+#include <array>
+#include <cmath>
```
Comment (aelovikov-intel) [inline sycl/include/sycl/types.hpp:85]: No, we cannot do that - `<cmath>` pollutes global namespace.

File: sycl/include/sycl/types.hpp [hunk 3]
```diff
@@ -571,9 +628,9 @@ template <typename Type, int NumElements> class vec {
       (Sz > detail::MaxVecAlignment);
 
   static constexpr bool IsHostHalf =
-      std::is_same_v<DataT, sycl::detail::half_impl::half> &&
-      std::is_same_v<sycl::detail::half_impl::StorageT,
-                     sycl::detail::host_half_impl::half>;
+      std::is_same<DataT, sycl::detail::half_impl::half>::value &&
+      std::is_same<sycl::detail::half_impl::StorageT,
+                   sycl::detail::host_half_impl::half>::value;
```
Comment (aelovikov-intel) [inline sycl/include/sycl/types.hpp:633]: Please revert this.

File: sycl/include/sycl/buffer.hpp (hunk 1 above)
Comment (cperkinsintel) [reply sycl/include/sycl/buffer.hpp:171]: The "old" sycl::vec implementation did not correctly support device_copyable for vec.  So this assert is only valid when using the "new" sycl::vec (ie, the -fpreview-breaking-changes flag).  There is no "else" as there was no assert before.

Comment (aelovikov-intel) [reply sycl/include/sycl/buffer.hpp:171]: https://github.com/intel/llvm/commit/68af512680ebd6de520fd54ab68847b23dedb209

Comment (cperkinsintel) [reply sycl/include/sycl/buffer.hpp:171]: Yes, I know.  But that PR was done recently building atop the "new" sycl:;vec, that we are now pushing backwards.  The "old" sycl::vec was not device_copyable (and not conformant) and cannot sustain that assertion.

File: sycl/include/sycl/types.hpp (hunk 3 above)
Comment (cperkinsintel) [reply sycl/include/sycl/types.hpp:633]: Good catch.

File: sycl/include/sycl/buffer.hpp (hunk 1 above)
Comment (cperkinsintel) [reply sycl/include/sycl/buffer.hpp:171]: nevermind - I'm being dumb as to what you are asking. Restored the assert we had previously.

File: sycl/include/sycl/types.hpp [hunk 4]
```diff
@@ -826,24 +905,24 @@ template <typename Type, int NumElements> class vec {
...
+          std::is_same<typename std::remove_const_t<Ty>, half>::value,
       vec &>
   operator=(const DataT &Rhs) {
     for (int i = 0; i < NumElements; ++i) {
       setValue(i, Rhs);
     }
     return *this;
   }
-#endif
+#endif // __SYCL_DEVICE_ONLY__
```
Comment (aelovikov-intel) [inline sycl/include/sycl/types.hpp:925]: This is definitely unrelated, same as other changes in this file.

**Comment by bader** [comment]
> This PR restores the old sycl::vec implementation and puts the new conforming one behind the flag/macro.

Technically, we added API versioning to handle "library-only" ABI changes like this. We should not need any flag/macro to change sycl::vec implementation.

File: sycl/include/sycl/types.hpp [hunk 5]
```diff
@@ -904,11 +983,21 @@ template <typename Type, int NumElements> class vec {
...
 
+  // TODO: Remove, for debug purposes only.
+  void dump() const {
+#ifndef __SYCL_DEVICE_ONLY__
+    for (int I = 0; I < NumElements; ++I) {
+      std::cout << "  " << I << ": " << getValue(I) << std::endl;
+    }
+    std::cout << std::endl;
+#endif // __SYCL_DEVICE_ONLY__
+  }
```
Comment (aelovikov-intel) [inline sycl/include/sycl/types.hpp:994]: Is this needed?

File: sycl/include/sycl/types.hpp [hunk 6]
```diff
@@ -904,11 +983,21 @@ template <typename Type, int NumElements> class vec {
...
+
 #ifdef __SYCL_DEVICE_ONLY__
   template <typename vector_t_ = vector_t,
-            typename =
-                typename std::enable_if_t<std::is_same_v<vector_t_, vector_t> &&
-                                          !std::is_same_v<vector_t_, DataT>>>
+            typename = typename std::enable_if_t<
+                std::is_same<vector_t_, vector_t>::value &&
+                !std::is_same<vector_t_, DataT>::value>>
```
Comment (aelovikov-intel) [inline sycl/include/sycl/types.hpp:1000]: This is still unrelated...

File: sycl/include/sycl/types.hpp [hunk 7]
```diff
@@ -925,8 +1014,143 @@ template <typename Type, int NumElements> class vec {
       return *ptr;
     }
   }
+#endif // __SYCL_DEVICE_ONLY__
+#else  // __INTEL_PREVIEW_BREAKING_CHANGES
```
Comment (aelovikov-intel) [inline sycl/include/sycl/types.hpp:1018]: I'd suggest to change all `ifdef`/`ifndef` to `#if defined()`/`#if !defined()` and update the comments to include `[!]defined` in `#else`/`#endif`.

File: sycl/include/sycl/types.hpp [hunk 8]
```diff
@@ -1076,13 +1300,15 @@ template <typename Type, int NumElements> class vec {
...
+#ifdef __INTEL_PREVIEW_BREAKING_CHANGES
+
 #ifdef __SYCL_DEVICE_ONLY__
 #define __SYCL_BINOP(BINOP, OPASSIGN, CONVERT)                                 \
   template <typename Ty = vec>                                                 \
   vec operator BINOP(const EnableIfNotUsingArrayOnDevice<Ty> &Rhs) const {     \
     vec Ret;                                                                   \
     Ret.m_Data = m_Data BINOP Rhs.m_Data;                                      \
-    if constexpr (std::is_same_v<Type, bool> && CONVERT) {                     \
+    if constexpr (std::is_same<Type, bool>::value && CONVERT) {                \
```
Comment (aelovikov-intel) [inline sycl/include/sycl/types.hpp:1311]: Still unrelated...

File: sycl/include/sycl/types.hpp [hunk 9]
```diff
@@ -1097,9 +1323,9 @@ template <typename Type, int NumElements> class vec {
   }                                                                            \
   template <typename T>                                                        \
   typename std::enable_if_t<                                                   \
-      std::is_convertible_v<DataT, T> &&                                       \
-          (std::is_fundamental_v<vec_data_t<T>> ||                             \
-           std::is_same_v<typename std::remove_const_t<T>, half>),             \
+      std::is_convertible<DataT, T>::value &&                                  \
+          (std::is_fundamental<vec_data_t<T>>::value ||                        \
+           std::is_same<typename std::remove_const_t<T>, half>::value),        \
```
Comment (aelovikov-intel) [inline sycl/include/sycl/types.hpp:1328]: Likewise.

**Review by aelovikov-intel — COMMENTED** [review]
I'd very much prefer to have the split at the headers level, instead of having `#if`s all the time. Maybe common part could be factored out to some helpers.

File: sycl/include/sycl/types.hpp [hunk 10]
```diff
@@ -925,8 +1014,143 @@ template <typename Type, int NumElements> class vec {
...
+  using EnableIfNotHostHalf = typename std::enable_if_t<
+      !std::is_same_v<DataT, sycl::detail::half_impl::half> ||
+          !std::is_same_v<sycl::detail::half_impl::StorageT,
+                          sycl::detail::host_half_impl::half>,
+      T>;
+  template <typename T = void>
+  using EnableIfHostHalf = typename std::enable_if_t<
+      std::is_same_v<DataT, sycl::detail::half_impl::half> &&
+          std::is_same_v<sycl::detail::half_impl::StorageT,
+                         sycl::detail::host_half_impl::half>,
```
[the remark this answers is not in the document]
Comment (cperkinsintel) [reply sycl/include/sycl/types.hpp:1031]: Your request is similar to the improvements I made when i originally wrote the "new" sycl::vec.  

But here, I'm simply restoring the "old" one. This is the code we had before. Everything in the #else of the PREVIEW_BREAKING_CHANGES macro is the restored old sycl::vec code. That's why the `::value` are showing up again, etc.  
At the next ABI-breaking window I want us to get rid of all of this old vec code once and for all.

File: sycl/include/sycl/types.hpp (hunk 5 above)
Comment (cperkinsintel) [reply sycl/include/sycl/types.hpp]: removed

File: sycl/include/sycl/types.hpp (hunk 7 above)
Comment (cperkinsintel) [reply sycl/include/sycl/types.hpp]: Based on our conversation, I've reorganized the files a bit, hopefully making it a bit easier to read. I've moved the vec implementations out to two files and added an additional injection file ( vec_gen_common.in ) that generates swizzles, binops and rellogops.

**Comment by aelovikov-intel** [comment]
+ @AlexeySachkov who is refactoring converts.
+ @steffenlarsen , just in case.

File: sycl/include/sycl/types.hpp [hunk 11]
```diff
@@ -10,2462 +10,8 @@
...
-  typename std::enable_if_t<!std::is_same_v<std::byte, T> &&
-                                (IsUsingArrayOnDevice || IsUsingArrayOnHost),
-                            vec<T, N>>
-  operator!() const {
-    vec Ret{};
-    for (size_t I = 0; I < NumElements; ++I)
-      Ret.setValue(I, !vec_data<DataT>::get(getValue(I)));
-    return Ret;
-  }
+#ifdef __INTEL_PREVIEW_BREAKING_CHANGES
```
Comment (aelovikov-intel) [inline sycl/include/sycl/types.hpp:13]: I think common parts of `vec_new`/`vec_old` at the start/end could be kept here still, outline only the middle piece with the big difference.

**Comment by cperkinsintel** [comment]
ping to reviewers. I'd like to get this merged before the next pulldown

**Comment by AlexeySachkov** [comment]
TL;DR I would suggest not to allow users to select between `vec` ABI versions to avoid having to bugfix both. Instead, I would suggest to use versioning mechanisms to support old `vec` so that the runtime works with binaries build by older toolchains, but always use new `vec` for any application built with newer runtime.

Details:

Haven't looked into this in details, but here are my two cents which are more about high-level design/approach/direction for this particular change and future changes like this we may have:

I'm concerned about multi-versioning: my `vec::convert` change has landed first, but there are anyway some regressions reported form it - does it mean that I will have to apply changes twice for both "new" and "old" implementation?

This could be inevitable in general case of having different API/ABI between versions and I think it would be helpful to have some comments/documentation somewhere which would instruct a developer about which versions require update depending on the changes, which version is actual, which version is legacy, etc.

Note: in this particular case it seems to me that we can actually unify both versions more or less nicely, because at first glance they seem similar: there is a difference of how underlying data type is chosen, but when it comes to operators implementation there seem to be just two implementations which are enough to cover all 4 combinations we have (legacy with/without ext-vec-type, new with/without ext-vec-type). 

I'm **not** asking to do that unification in scope of this PR, because it will require notable efforts and may not work out as good as I imagine. Plus, we could use this PR to troubleshoot our versioning mechanism we have:

>> This PR restores the old sycl::vec implementation and puts the new conforming one behind the flag/macro.
> Technically, we added API versioning to handle "library-only" ABI changes like this. We should not need any flag/macro to change `sycl::vec` implementation.

This is a very good point and I would like to discuss it more. Do we have a use case where uses would actually need a control over `vec` ABI? Something like interoperability between object files compiled by different toolchain versions? Is it really common?

If we change our mindset to "`new` implementation is a current implementation, this is what we will have in new releases", then my previous question about "do I need to change both when I fix a bug" is a no question at all. "old" is only for backward compatibility, no new releases will be done with it and therefore no need to touch it at all. I only need to fix a new one for the upcoming release.

We would still have to maintain some `#ifdef __breaking_changes` macro because of API breaks we are not allowed to make which are required to do API bugfix or to match newer SYCL specs.

However, I doubt that this is a header-only ABI break, because there are host versions of math built-ins, which are implemented within the library and therefore we may have to duplicate all (?) those library entry points to maintain backward compatibility with older runtime. Will it cause even more code changes? Likely so. Will it increase maintenance burden more than having to `vec` implementations? I don't think so, because I expect `vec` interface or implementation to be changed/bugfixed more often that implementations of math built-ins for host.

File: sycl/include/sycl/types.hpp [hunk 12]
```diff
@@ -10,2207 +10,8 @@
...
```
Comment (AlexeySachkov) [inline sycl/include/sycl/types.hpp:13]: I would vote for adding a comment here about how to proceed with bugfix/further changes to `vec` class. If I understand correctly, all new API/ABI breaking changes should only go into `vec_new`, whilst generic implementation bugfix should be done in both versions

**Review by AlexeySachkov — APPROVED** [review]
We had a discussion about this and we are not sure that we can be safe about passing SYCL-CTS with the old implementation. Considering other constraints we have, the decision was (if I understood it correctly) to proceed with this patch as-is as the only available option at the moment. Therefore, approving the PR (even though I'm not really a code owner for this code)

**Comment by aelovikov-intel** [comment]
> I'd very much prefer to have the split at the headers level, instead of having `#if`s all the time. Maybe common part could be factored out to some helpers.

I see that it was a bad suggestion now... How hard would it be to return it back?

File: sycl/include/sycl/types.hpp [hunk 13]
```diff
@@ -1205,6 +1497,19 @@ template <class T, class... U,
...
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+// This file generates swizzles, binops and rellogops for sycl::vec.
+// It is used by injection ( direct #include into body ) from both vec_new.hpp
+// or vec_old.hpp and should not be used by anything else.
+//
+
```
Comment (aelovikov-intel) [inline sycl/include/sycl/types.hpp:1512]: Drop this.

File: sycl/include/sycl/types.hpp [hunk 14]
```diff
@@ -1992,45 +2319,60 @@ __SYCL_DEFINE_VECSTORAGE_IMPL_FOR_TYPE(double, double)
 
 #undef __SYCL_DEFINE_VECSTORAGE_IMPL_FOR_TYPE
 #undef __SYCL_DEFINE_VECSTORAGE_IMPL
-#endif // __SYCL_DEVICE_ONLY__
+#endif // ndef __SYCL_USE_EXT_VECTOR_TYPE__
```
Comment (aelovikov-intel) [inline sycl/include/sycl/types.hpp:2322]: typo

**Review by aelovikov-intel — APPROVED** [review]
LGTM given two issues with comments will be fixed before merge.
